### PR TITLE
F 84 provide reasons to deactivating project

### PR DESCRIPTION
**Pull request by mohammadranjbarz**
related to Giveth/giveth-io-typescript#84

What Id did on this PR:
1. Create a table named `project_status_reason` for adding template reasons there https://github.com/Giveth/giveth-io-typescript/issues/84#issuecomment-1016941722
2. Add this entity to adminBro so admins can add new reasons
3. Add a new table for adding history changes there 
4. Add new record when someone activate/deactivate from our dapp ( ideally in future we should add record even if admin change status from adminBro)
5. Add projectStatusHistory to admin bro, so admins can see the changes and history of a project there
6. Refactor deactivate/activate mutations
7. Writing some test cases

PS: @CarlosQ96  I wrote down above items to make the review easy for you

### CODE REVIEW

File: migration/1643957749780-addProjectStatusReasonTable.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+                  CONSTRAINT "FK_2edf0fca3c1aca99d1c4924e6a2" FOREIGN KEY ("statusId")
+                      REFERENCES public.project_status (id) MATCH SIMPLE
+              )
+`,
+    );
+  }
+
+  async down(queryRunner: QueryRunner): Promise<void> {
+    try {
+      await queryRunner.query(`DROP TABLE "project_status_reason"`);
```
Comment (CarlosQ96) [inline migration/1643957749780-addProjectStatusReasonTable.ts]: Remember if you rollback and run the server, synchronization will create the entity schema again. 

Second scenario would be if the person deletes the entity, run server and then rollbacks. It will break the migrations as sync mode already deleted the table.

I think we can delegate this to the sync mode and leave a comment.
What do you think?
Maybe we should document how to rollback correctly. But you can decide if it's fine for now or not.

File: test/testUtils.ts [hunk 2]
```diff
@@ -208,6 +210,24 @@ export const SEED_DATA = {
     'food7',
     'food8',
   ],
+  STATUS_REASONS: [
+    {
+      description: 'The project has completed its goals!',
+      statusId: ProjStatus.deactive,
+    },
+    {
+      description: 'The project is not longer active.',
```
Comment (CarlosQ96) [inline test/testUtils.ts]: Just a typo. It should should be "no" instead of "not".
"The project is no longer active."

File: migration/1643962364050-seedProjectStatusreasons.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+export class seedProjectStatusreasons1643962364050
+  implements MigrationInterface
+{
+  async up(queryRunner: QueryRunner): Promise<void> {
+    // https://github.com/Giveth/giveth-io-typescript/issues/84#issuecomment-1016941722
+    await queryRunner.query(`INSERT INTO public.project_status_reason ("statusId", description ) VALUES 
+                    (${ProjStatus.deactive},'The project has completed its goals!'),
+                    (${ProjStatus.deactive},'The project is no longer in need of funding.'),
+                    (${ProjStatus.deactive},'The project is not longer active.'),
```
Comment (CarlosQ96) [inline migration/1643962364050-seedProjectStatusreasons.ts]: Just a small typo. It should be "no" instead of "not".
"The project is no longer active."

File: src/resolvers/projectResolver.ts [hunk 4]
```diff
@@ -1141,7 +1144,26 @@ export class ProjectResolver {
...
+      };
+      analytics.track(
+        SegmentEvents.PROJECT_ACTIVATED,
+        `givethId-${ctx.req.user.userId}`,
+        segmentProject,
+        null,
+      );
+      return true;
     } catch (error) {
       SentryLogger.captureException(error);
```
Comment (CarlosQ96) [inline src/resolvers/projectResolver.ts:1169]: we should add the error in our logs like you did in the deactivate mutation, to keep consistency.
      logger.error('projectResolver.activateProject() error', error);

File: src/entities/project.ts [hunk 5]
```diff
@@ -341,13 +351,38 @@ class Project extends BaseEntity {
...
   // Status 7 is deleted status
   mayUpdateStatus(user: User) {
-    if (this.statusId === ProjStatus.cancel) return false;
+    if (this.statusId === ProjStatus.cancel) {
+      throw new Error(
+        errorMessages.THIS_PROJECT_IS_CANCELLED_OR_DEACTIVATED_ALREADY,
+      );
+    }
 
     if (this.users.filter(o => o.id === user.id).length > 0) {
```
Comment (CarlosQ96) [inline src/entities/project.ts]: I have a faint memory this works fine now, but on really  REALLY REALLLY Oold projects won't. It seems they don't have the user relation set up correctly or non existant, but they have admin string. This happened months ago.

I think we should use the filter as is but also add an OR  for checking the admin string which is also important.
```
if (this.users.filter(o => o.id === user.id).length > 0 || user.id === Number(this.admin))
```

**Review by CarlosQ96 — APPROVED** [review]
I ran the tests, queries and checked all the code. It's seems working good! Great job Mohammad!
I left some comment for some minor things and one edge case we encountered like easily 6 months ago.

File: src/entities/project.ts (hunk 5 above)
Comment (CarlosQ96) [reply src/entities/project.ts:360]: Or just the admin string... which honestly has the source of truth, without filtering a list of users.

File: migration/1643962364050-seedProjectStatusreasons.ts (hunk 3 above)
Comment (mohammadranjbarz) [reply migration/1643962364050-seedProjectStatusreasons.ts]: Oh thanks it has been fixed

File: src/resolvers/projectResolver.ts (hunk 4 above)
Comment (mohammadranjbarz) [reply src/resolvers/projectResolver.ts:1169]: You are exactly write, I added the logger

File: src/entities/project.ts (hunk 5 above)
Comment (mohammadranjbarz) [reply src/entities/project.ts]: Nice catch, I have changed the condition

File: test/testUtils.ts (hunk 2 above)
Comment (mohammadranjbarz) [reply test/testUtils.ts]: Thanks, changed

File: migration/1643957749780-addProjectStatusReasonTable.ts (hunk 1 above)
Comment (mohammadranjbarz) [reply migration/1643957749780-addProjectStatusReasonTable.ts]: You are right, I deleted them
